### PR TITLE
Build Ubuntu CI containers on GitHub Actions

### LINKED ISSUE #74: Refresh ci-debian-buster container image used for build
Followon from open-quantum-safe/liboqs#1702

The openquantumsafe/ci-debian-custer-amd64:latest image is used as part of our ci process

When working on the above PR I noticed we weren't pinning the version spec of this image (though the tooling did not detect this).

I inspected the image with a [scan on quay.io](https://quay.io/repository/planetf1/ci-debian-buster-amd64/manifest/sha256:b44ec894a0071432ca993aac617e2e503d03984f2079eee48a98a136fe0f0dbc?tab=vulnerabilities)

This is used for testing/verification, rather than supplying images for consumers, but it looks as if it could do with being updated - any images/sw used for tests could be compromised to hide an injected vulnerability.

The current image has quite old java (1.11.0), and also older versions of qemu and other tools. [See list](https://quay.io/repository/planetf1/ci-debian-buster-amd64/manifest/sha256:b44ec894a0071432ca993aac617e2e503d03984f2079eee48a98a136fe0f0dbc?tab=packages)

### LINKED ISSUE #81: Update Container build CI
The current CI has the following problems:
- Still uses CircleCI that we want to get away from
- Still requires docker login during PR which excludes external contributors from running CI
- Still only builds single-platform images instead of multi-platform ones

This issue is to suggest fixing these shortcomings before/while upgrading to ubuntu:latest as per https://github.com/open-quantum-safe/liboqs/issues/1780

### PR DESCRIPTION
**Pull request by SWilson4**
This PR attempts to accomplish the following:
- Move our CI image build logic from CircleCI to GitHub Actions,
- Stop building images for Linux distributions not listed as [supported platforms for `liboqs`](https://github.com/open-quantum-safe/liboqs/blob/66f713f/PLATFORMS.md),
- Start building images for Ubuntu Jammy and Noble on arm64, and
- Rework our CI tagging system to use multiarch images.

Fixes #81, fixes #74.

Related issues: https://github.com/open-quantum-safe/liboqs/issues/1780.

I'm leaving this as draft until the necessary environment variables/secrets are added (cc @ryjones). And, until it's not a Friday, since this could break all of our CI, and I'd rather not deal with that on the weekend. :)

I may need to further update these Dockerfiles (e.g., to add a dependency) once I have a chance to use them in liboqs CI while working on https://github.com/open-quantum-safe/liboqs/issues/1780.

### CODE REVIEW

**Comment by ryjones** [comment]
@SWilson4 looks like I need to make a dockerhub token? anything else?

**Review by dstebila — APPROVED** [review]

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: Build Docker images
+
+on:
+  pull_request:
+  push:
```
Comment (baentsch) [inline .github/workflows/build.yml:5]: Does this mean again double execution of all tasks (once at push, then again at PR)?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+jobs:
+  ubuntu:
+    strategy:
+      matrix:
+        arch:
+          - arm64
+          - x86_64
+        distro:
+          - focal
+          - jammy
```
Comment (baentsch) [inline .github/workflows/build.yml:17]: Why Jammy? It's not listed in https://github.com/open-quantum-safe/liboqs/blob/main/PLATFORMS.md
Focal is listed, but shouldn't be (if this lands) -- so remove this once PLATFORMS gets updated?

File: .github/workflows/push.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      matrix:
+        distro:
+          - focal
+          - jammy
+          - latest
+    runs-on: oqs-arm64
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Login to Docker Hub
```
Comment (baentsch) [inline .github/workflows/push.yml:19]: Is this needed on a non-pushing action?

**Review by baentsch — APPROVED** [review]
Logically LGTM; some nits and questions on efficiency.

File: .github/workflows/build.yml (hunk 2 above)
Comment (SWilson4) [reply .github/workflows/build.yml:17]: My intent was to update PLATFORMS.md to list support for Focal, Jammy, and Noble (i.e., every Ubuntu LTS version still receiving [standard security maintainence](https://ubuntu.com/about/release-cycle)) on x86_64 and arm64. Do we really want to support only the latest LTS? Focal will be receiving security updates for another year and Jammy for another three.

File: .github/workflows/push.yml (hunk 3 above)
Comment (SWilson4) [reply .github/workflows/push.yml:19]: This is a pushing action that will only run on `main`.

File: ubuntu-focal/Dockerfile [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ARG ARCH
-FROM multiarch/ubuntu-core:${ARCH}-focal
-LABEL version="4"
+FROM ubuntu:focal
```
[the remark this answers is not in the document]
Comment (SWilson4) [reply ubuntu-focal/Dockerfile:2]: It's used later in the file to decide whether or not to install Doxygen.

File: .github/workflows/build.yml (hunk 1 above)
Comment (SWilson4) [reply .github/workflows/build.yml:5]: Yes, unfortunately. :( I would be happy to adopt @praveksharma's suggestion for [reducing this](https://github.com/open-quantum-safe/liboqs/pull/1902#issuecomment-2310264472), though.

As an aside, I looked into the duplicate run issue [a while time ago](https://github.com/open-quantum-safe/tsc/issues/5#issuecomment-2029868313) and didn't find any automated solution that I was happy with.

**Comment by SWilson4** [comment]
> @SWilson4 looks like I need to make a dockerhub token? anything else?

It would be nice but not necessary to have the dockerhub username (oqsbot) saved in a non-secret variable. Other than that nothing.

File: .github/workflows/build.yml (hunk 1 above)
Comment (ryjones) [reply .github/workflows/build.yml:5]: you could limit the push to main branch

Comment (baentsch) [reply .github/workflows/build.yml:5]: OK here -- this stuff doesn't run as often as code changes in `liboqs`...

File: .github/workflows/push.yml (hunk 3 above)
Comment (baentsch) [reply .github/workflows/push.yml:19]: Sorry for the noise: I was convinced I saw and commented on this in the "non-pushing" build job: In the final push it obviously is needed.

File: ubuntu-focal/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (baentsch) [reply ubuntu-focal/Dockerfile:2]: Ah, focal had an overly old doxygen built-in that needed replacement, right. Thanks for the reminder.

**Comment by ryjones** [comment]
> > @SWilson4 looks like I need to make a dockerhub token? anything else?
> 
> It would be nice but not necessary to have the dockerhub username (oqsbot) saved in a non-secret variable. Other than that nothing.

Done. Do you control the oqsbot user on dockerhub?

**Comment by ryjones** [comment]
@SWilson4 I've added `DOCKERHUB_TOKEN` as an org secret, and granted access only to this repo, for now.

**Review by praveksharma — APPROVED** [review]

**Comment by baentsch** [comment]
> @SWilson4 I've added `DOCKERHUB_TOKEN` as an org secret, and granted access only to this repo, for now.

What does it take to also add it to oqs-demos, @ryjones ? Would allow implementing https://github.com/open-quantum-safe/oqs-demos/issues/294

**Comment by baentsch** [comment]
> > @SWilson4 I've added `DOCKERHUB_TOKEN` as an org secret, and granted access only to this repo, for now.
> 
> What does it take to also add it to oqs-demos, @ryjones ? Would allow implementing [open-quantum-safe/oqs-demos#294](https://github.com/open-quantum-safe/oqs-demos/issues/294)

Disregard. It already works there -- just wonder why now :-)

File: .github/workflows/build.yml (hunk 2 above)
Comment (SWilson4) [reply .github/workflows/build.yml:17]: That seems reasonable to me.

**Comment by SWilson4** [comment]
I'm going to merge now so that I can try out these images in liboqs CI. I may need to add additional dependencies in a follow-up PR.

**Comment by SWilson4** [comment]
This failed at merge due to an incorrect Docker Hub login. :( Could you please take a look @ryjones?

**Comment by ryjones** [comment]
@SWilson4 fixed :)

**Pull request merged**